tokio-net/src/driver/sharded_slab/mod.rs

```diff
@@ -0,0 +1,307 @@
+//! A lock-free concurrent slab.
+mod page;
+mod tid;
+pub(crate) use tid::Tid;
+mod iter;
+
+use crate::sync::{
+    atomic::{AtomicUsize, Ordering},
+    CausalCell,
+};
+use page::{slot, Addr, Page};
+use std::fmt;
+
+#[cfg(target_pointer_width = "64")]
+const MAX_THREADS: usize = 4096;
+#[cfg(target_pointer_width = "32")]
+const MAX_THREADS: usize = 2048;
+
+const MAX_PAGES: usize = 16;
+const RESERVED_BITS: usize = 5;
+
+/// A sharded, lock-free slab.
+pub(crate) struct Slab<T> {
+    shards: Box<[CausalCell<Shard<T>>]>,
+}
+
+struct Shard<T> {
+    tid: usize,
+    sz: usize,
+    len: AtomicUsize,
+    // ┌─────────────┐      ┌────────┐
+    // │ page 1      │      │        │
+    // ├─────────────┤ ┌───▶│  next──┼─┐
+    // │ page 2      │ │    ├────────┤ │
+    // │             │ │    │XXXXXXXX│ │
+    // │ local_free──┼─┘    ├────────┤ │
+    // │ global_free─┼─┐    │        │◀┘
+    // ├─────────────┤ └───▶│  next──┼─┐
+    // │   page 3    │      ├────────┤ │
+    // └─────────────┘      │XXXXXXXX│ │
+    //       ...            ├────────┤ │
+    // ┌─────────────┐      │XXXXXXXX│ │
+    // │ page n      │      ├────────┤ │
+    // └─────────────┘      │        │◀┘
+    //                      │  next──┼───▶
+    //                      ├────────┤
+    //                      │XXXXXXXX│
+    //                      └────────┘
+    //                         ...
+    pages: Vec<Page<T>>,
```

Comment:
Similarly to https://github.com/tokio-rs/tokio/pull/1625/files#r330761039, this _could_ become a fixed array.